### PR TITLE
[eBPF] Let deepflow-agent run on arm architecture (#1035)

### PR DESCRIPTION
**Pull request by yinjiping**
Let deepflow-agent run on arm architecture 

### This PR is for:

- Agent

#### Affected branches
- main

### CODE REVIEW

File: agent/docker/DockerfileToFix-aarch64.build [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# 编译阶段
+# 使用 Centos 为基础镜像，安装所需依赖
+# aarch64 agent 编译镜像，需要在aarch64的机器中制作此镜像
```
Comment (sharang) [inline agent/docker/DockerfileToFix-aarch64.build:3]: 用英文注释吧

File: agent/docker/DockerfileToFix-aarch64.build [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+# 2. 配置rust环境变量
+ENV RUSTUP_HOME=/usr/local/rustup \
+    CARGO_HOME=/usr/local/cargo \
+    PATH=/usr/local/cargo/bin:$PATH \
+    http_proxy=http://47.243.232.26:8888 \
+    https_proxy=http://47.243.232.26:8888 \
+    no_proxy=localhost,127.0.0.1,localaddress,.localdomain.com,192.168.0.0/16,10.0.0.0/8,172.16.0.0/12,100.64.0.0/10,17.0.0.0/8,.local,169.254.0.0/16,224.0.0.0/4,240.0.0.0/4,.yunshan.net,.cn,aliyun.com,aliyuncs.com
+
+
+# 安装基础依赖#
```
Comment (sharang) [inline agent/docker/DockerfileToFix-aarch64.build:15]: 去掉末尾#

File: agent/docker/DockerfileToFix-aarch64.build [hunk 3]
```diff
@@ -0,0 +1,71 @@
+# 编译阶段
+# 使用 Centos 为基础镜像，安装所需依赖
+# aarch64 agent 编译镜像，需要在aarch64的机器中制作此镜像
+FROM centos:centos7.9.2009
+
+# 2. 配置rust环境变量
+ENV RUSTUP_HOME=/usr/local/rustup \
+    CARGO_HOME=/usr/local/cargo \
+    PATH=/usr/local/cargo/bin:$PATH \
+    http_proxy=http://47.243.232.26:8888 \
```
Comment (nrjatyunshan) [inline agent/docker/DockerfileToFix-aarch64.build:10]: 这是个公网IP的代理地址,暴露出来会不会有问题

Comment (mytting) [reply agent/docker/DockerfileToFix-aarch64.build:10]: 在github不需要代理了，去掉吧，参考开源的x86 的这个dockerfile

**Comment by mytting** [comment]
@sharang @yinjiping 阿里云海外/香港等等region都不支持arm实例，aws 支持，但是之前调研过的k8s方案有些问题
- eks集群不支持arm架构的spot
- 存储方面有个功能不支持
ec2方案有些复杂，难度有些高

约了aws的工程师明天搞一个骚操作试一下，不可行的话就要用ec2方案了

**Comment by aktech** [comment]
Hey guys, let me know if you need any help with [cirun.io](https://cirun.io) solution. It supports arm based machines too.

**Comment by mytting** [comment]
> Hey guys, let me know if you need any help with [cirun.io](https://cirun.io) solution. It supports arm based machines too.

Thanks for your attention, [cirun.io](https://cirun.io) is a very good tool, but I found that the waiting time of action job is too long, maybe more than 5 minutes, which leads to the high CI overall time

**Comment by aktech** [comment]
> Thanks for your attention, [cirun.io](https://cirun.io/) is a very good tool, but I found that the waiting time of action job is too long, maybe more than 5 minutes, which leads to the high CI overall time

Hey @mytting Although we're are working on making it faster, we're bottlenecked by cloud providers mostly, so unfortunately it is true at this point in time.

**Comment by mytting** [comment]
> > Thanks for your attention, [cirun.io](https://cirun.io/) is a very good tool, but I found that the waiting time of action job is too long, maybe more than 5 minutes, which leads to the high CI overall time
> 
> Hey @mytting Although we're are working on making it faster, we're bottlenecked by cloud providers mostly, so unfortunately it is true at this point in time.

Maybe there is something wrong with my machine_image. It seems that it is not possible for me to use the official image of aws directly, so it is OK for me to use the image provided by you

**Comment by aktech** [comment]
Also, I am happy to help support your use case and do what I can to make it worthwhile for you guys.

**Comment by aktech** [comment]
>Maybe there is something wrong with my machine_image. It seems that it is not possible for me to use the official image of aws directly, so it is OK for me to use the image provided by you

It might just be needed to be subscribed to a particular image if you're using an AWS Marketplace image.

Which image is causing issues and and in which region?

**Comment by mytting** [comment]
> > Maybe there is something wrong with my machine_image. It seems that it is not possible for me to use the official image of aws directly, so it is OK for me to use the image provided by you
> 
> It might just be needed to be subscribed to a particular image if you're using an AWS Marketplace image.
> 
> Which image is causing issues and and in which region?
> 
> Ideally you should be able to use any image, we don't block or restrict that.

Such as public image: ami fd8802f94ed1c969 0, region: us-east-2

**Comment by mytting** [comment]
> > Maybe there is something wrong with my machine_image. It seems that it is not possible for me to use the official image of aws directly, so it is OK for me to use the image provided by you
> 
> It might just be needed to be subscribed to a particular image if you're using an AWS Marketplace image.
> 
> Which image is causing issues and and in which region?
> 
> Ideally you should be able to use any image, we don't block or restrict that.

When I use the image in your instance and specify the region, I can pull up a runner within two minutes, which is within our acceptable range. I want to know which images are available

**Comment by aktech** [comment]
> Such as public image: ami fd8802f94ed1c969 0, region: us-east-2

Are you sure, this image exists in `us-east-2`? I couldn't find it:
Screenshot 2022-10-16 at 9 28 14 pm

> When I use the image in your instance and specify the region, I can pull up a runner within two minutes, which is within our acceptable range. I want to know which images are available

So, you can use any image from public or marketplace or even create your own image. I can debug if I can find the image you mentioned.

Are you available for a quick call to debug this?

**Comment by mytting** [comment]
> > 如公共镜像：ami fd8802f94ed1c969 0，地区：us-east-2
> 
> 你确定，这张图片存在于`us-east-2`？我找不到它： 截图 2022-10-16 at 9 28 14 pm
> 
> > 当我在你的实例中使用图像并指定区域时，我可以在两分钟内拉起一个跑步者，这在我们可以接受的范围内。我想知道有哪些图片可用
> 
> 因此，您可以使用来自公共或市场的任何图像，甚至可以创建自己的图像。如果我能找到你提到的图像，我可以调试。
> 
> 您可以快速致电调试吗？

ami-0149b2da6ceec4bb0

**Comment by aktech** [comment]
> ami-0149b2da6ceec4bb0

Thanks, let me take a look.

**Comment by mytting** [comment]
> > ami-0149b2da6ceec4bb0
> 
> Thanks, let me take a look.

image
This is the

**Comment by aktech** [comment]
This image (`ami-0149b2da6ceec4bb0`) is in `us-east-1` and not `-2`.  You need to specify the correct region in the `.cirun.yml` file in order to get the runner.

I just created a runner with this AMI here: https://github.com/aktech/cirun-sandbox/actions/runs/3260895245/jobs/5354884188

With this `cirun.yml` configuration:

https://github.com/aktech/cirun-sandbox/blob/ba41ba1958bbdc5cd45655045398d23bad7f04b2/.cirun.yml

```yml
runners:
    - name: aws-runner
      cloud: aws
      instance_type: t2.large
      region: us-east-1
      machine_image: ami-0149b2da6ceec4bb0
      labels:
        - cirunner
```

**Comment by aktech** [comment]
Quick Tip: You can also see the errors with runner creation if you click on the Details link Next to **Cirun.io** here:

---
Screenshot 2022-10-16 at 10 04 36 pm

**Comment by mytting** [comment]
> Quick Tip: You can also see the errors with runner creation if you click on the Details link Next to **Cirun.io** here:

Ok, thank you, I will try again, maybe my quota is relatively small, now can only 32C, can not concurrency

**Comment by mytting** [comment]
> Quick Tip: You can also see the errors with runner creation if you click on the Details link Next to **Cirun.io** here:
> 
> Screenshot 2022-10-16 at 10 04 36 pm

Also, how long will it take for the default Idled to delete the runner and machine?

**Comment by mytting** [comment]
I found that after I cancelled the job, I would create a runner and clean it up automatically when there were no tasks to execute, which might cause waste
@aktech

**Comment by aktech** [comment]
> Ok, thank you, I will try again, maybe my quota is relatively small, now can only 32C, can not concurrency

Yeah, it's usually limited initially, you can request AWS, they will increase the quota.

> Also, how long will it take for the default Idled to delete the runner and machine?

It's pretty much immediately, when the workflow completes, irrespective of failure or success. If by any chance it isn't, the idle runners are cleaned up regularly every 30 minutes. Let me know if it didn't, I don't expect to have lingering resources left by Cirun.

**Comment by mytting** [comment]
> > Ok, thank you, I will try again, maybe my quota is relatively small, now can only 32C, can not concurrency
> 
> Yeah, it's usually limited initially, you can request AWS, they will increase the quota.
> 
> > Also, how long will it take for the default Idled to delete the runner and machine?
> 
> It's pretty much immediately, when the workflow completes, irrespective of failure or success. If by any chance it isn't, the idle runners are cleaned up regularly every 30 minutes. Let me know if it didn't, I don't expect to have lingering resources left by Cirun.

image
Here is a dirty data, there is actually no vm, and can the cleaning interval be configured? 30 minutes might be a little long

**Comment by aktech** [comment]
> Here is a dirty data, there is actually no vm, and can the cleaning interval be configured? 30 minutes might be a little long

I'll take a look, might be a bug. Cleaning interval is actually 4 minutes, but it cleans runners that are ideal for 15 minutes (not 30 my bad). Also, this is just a fail-safe mechanism, generally VMs are cleared immediately when we receive webhook event from GitHub, excluding some exceptional cases.

**Comment by aktech** [comment]
> @aktech How do I configure the vm disk size?

Currently the only way to do that is to create a custom AMI with larger disk size:
https://docs.aws.amazon.com/toolkit-for-visual-studio/latest/user-guide/tkv-create-ami-from-instance.html

**Comment by yinjiping** [comment]
重复的PR

**Pull request closed without merging**